src/algorithms/genetic.rs

```diff
@@ -0,0 +1,376 @@
+
+    for i in neighbors {
+        if solution[node] == solution[i] {
+            is_valid = false;
+        }
+    }
+
+    is_valid
+}
+
+// A coloring upper bound based on the largest neighborhood
+// Tighter upper bounds helps a lot in the randomized color
+// assignment (individual generation and mutation)
+/// Determine an upper bound for the number of colors in a graph.
+///
+/// This upper bound is calculated based on the `Brook's theorem` (i.e., the chromatic number is less or equal than the maximum vertex degree of the graph plus one).
+fn coloring_upper_bound(graph: &Graph) -> usize {
+    let mut colors = 0;
+
+    for i in 0..graph.num_vertices() {
+        let neighborhood_size = graph.get_neighbors(i).len();
+
+        if neighborhood_size + 1 > colors {
+            colors = neighborhood_size + 1;
+        }
+    }
+
+    colors
+}
+
+/// Generates a randomized solution for the GCP problem.
+///
+/// A solution consists of a vector of size `n`, with `n` being the number of vertices of the graph,
+/// where the `ith position` receives a number between `1` and the `upper bound for coloring`.
+/// This number represents the `color of the ith vertex` in the current solution.
+fn generate_individual(graph: &Graph, upper_bound: usize) -> Vec<usize> {
+    let n = graph.num_vertices();
+    let mut individual = vec![0; n];
+    let mut is_invalid = true;
+
+    for i in 0..n {
+        while is_invalid {
+            individual[i] = rand::thread_rng().gen_range(1..=upper_bound);
+            is_invalid = !valid_color_assignment(&graph, &individual, i);
+        }
+
+        is_invalid = true;
+    }
+
+    individual
+}
+
+/// Traverses the solution vector, changing the color of each vertex to a random color
+/// with a probability given by the `mutation_probability` parameter.
+fn mutate(
+    graph: &Graph,
+    individual: &mut Vec<usize>,
+    upper_bound: usize,
+    mutation_probability: &f64,
+) {
+    let n = graph.num_vertices();
+    let mut rng = rand::thread_rng();
+    let mut is_invalid = true;
+
+    for i in 1..n {
+        let rand = rng.gen_range(0.0..1.0);
+
+        if rand <= *mutation_probability {
+            while is_invalid {
+                individual[i] = rng.gen_range(1..=upper_bound);
+                is_invalid = !valid_color_assignment(&graph, &individual, i);
+            }
+
+            is_invalid = true;
+        }
+    }
+}
+
+/// Selects 20% of the fittest individuals in the population and, from them, randomly
+/// selects two individuals who will be the parents of an offspring.
+fn select(
+    population: &Vec<(usize, Vec<usize>)>,
+    population_size: usize,
+) -> (Vec<usize>, Vec<usize>) {
+    let mut rng = rand::thread_rng();
+    let limit = (population_size as f64 * 0.2).floor() as usize;
+
+    let p1 = rng.gen_range(0..=limit);
+    let mut p2 = rng.gen_range(0..=limit);
+
+    while p1 == p2 {
+        p2 = rng.gen_range(0..=limit);
+    }
+
```

Comment:
Tem um método `choose` na crate `rand` pra você não precisar lidar com isso

Reply:
Acho que isso aqui deve resolver:
```suggestion
let limit = (population_size as f64 * 0.2).floor() as usize;
let mut best_pop = population.clone();
best_pop.truncate(limit + 1);
let p: Vec<(usize, Vec<usize>)> = best_pop.choose_multiple(&mut rng, 2).cloned().collect();
(p[0].1.clone(), p[1].1.clone())
```

Reply:
Ainda melhor (mas confuso):

```suggestion
    let (_, mut colors): (Vec<usize>, Vec<Vec<usize>>) =
        population.to_owned().clone().into_iter().unzip();

    colors.truncate(limit + 1);

    let p: Vec<Vec<usize>> = colors
        .choose_multiple(&mut rand::thread_rng(), 2)
        .cloned()
        .collect();
    (p[0].clone(), p[1].clone())

```